src/randomeffectsterm.jl

```diff
@@ -49,15 +22,40 @@ function StatsModels.termvars(t::RandomEffectsTerm)
     vcat(StatsModels.termvars(t.lhs), StatsModels.termvars(t.rhs))
 end
 
+StatsModels.terms(t::RandomEffectsTerm) = union(StatsModels.terms(t.lhs), StatsModels.terms(t.rhs))
+
+# | in MixedModel formula -> RandomEffectsTerm
 function StatsModels.apply_schema(
     t::FunctionTerm{typeof(|)},
     schema::MultiSchema{StatsModels.FullRank},
     Mod::Type{<:MixedModel}
 )
     lhs, rhs = t.args_parsed
 
+    isempty(intersect(StatsModels.termvars(lhs), StatsModels.termvars(rhs))) ||
+        throw(ArgumentError("Same variable appears on both sides of |"))
+
+    apply_schema(RandomEffectsTerm(lhs, rhs), schema, Mod)
+end
+
+# allowed types (or tuple thereof) for blocking variables (RHS of |):
+const GROUPING_TYPE = Union{<:CategoricalTerm, <:InteractionTerm{<:NTuple{N,CategoricalTerm} where {N}}}
+check_re_group_type(term::GROUPING_TYPE) = true
+check_re_group_type(terms::Tuple{Vararg{<:GROUPING_TYPE}}) = true
+check_re_group_type(x) = false
```

Comment:
should we explicitly mark these as private/internal with `_`?

Reply:
they're not exported so I dunno, up to you I guess.

Reply:
Hmmm.... Not exported, not advertised and lacking a docstring. People should get the hint. :smile: